src/WebJobs.Script/HostNameProvider.cs

```diff
@@ -32,7 +35,7 @@ public virtual string Value
         {
             get
             {
-                if (string.IsNullOrEmpty(_hostName))
+                if (!_isSet && string.IsNullOrEmpty(_hostName))
```

Comment:
We need to check whether this is safe w.r.t. specialization. It was coded conservatively like this to allow for the possibility that specialization may change these environment variables.

Reply:
Indeed - we can go either way here. In this case, we're registered on the ScriptHost, so a new instance will be created on specialization here and there should be no downsides to caching (I've tried to ensure this with all PRs but forgot to comment on this one).

I'd love to discuss environmental variables in general since they add up from 1-3% of our cost - I wonder if we can cache _all_ of them, and flush that cache entirely when _any_ change (a relatively rare event vs. access), that would make a lot of these code paths faster, simpler, and easier to test. It could also add consistency where we sometimes have settings and sometimes have environmental variables today, if we want that layering (totally optional). As a new user setting up tests and trying to get a many green as possible, that's an area I hit some duplication/friction and have been wondering about it a bit.

Reply:
@NickCraver as discussed, let's open an issue for the environment variables caching so we can explore that. We used to do a lot of that and beyond the perf gains, I agree it would help drive consistency.

Reply:
:+1: - had some StackExchange.Redis items in queue but will try and get that written up by tomorrow, going to close this one on out!

Reply:
I took the time to prototype the proposal to illustrate some thinking a bit better on the consumption side - issue written up in #8098 :)